publish.go

```diff
@@ -0,0 +1,326 @@
+	log.Infof("New version codes: '%v'", newVersionCodes)
+
+	var cleanTrack bool
+	if len(release.VersionCodes) != len(newVersionCodes) {
+		log.Warnf("Mismatching app count, removing (%v) versions from release: %s", release.VersionCodes, release.Name)
+		cleanTrack = true
+	} else {
+		log.Debugf("The number of App version codes (current and new) are equal")
+		release.VersionCodes = sortAndFilterVersionCodes(release.VersionCodes, newVersionCodes)
+	}
+
+	if cleanTrack {
+		log.Infof("Clearing version codes for release %v, the new versions will be %v", release.Name, newVersionCodes)
+		release.VersionCodes = newVersionCodes
+	}
+}
+
+// sortAndFilterVersionCodes sorts and filters two set of version codes, returns the higher for each.
+func sortAndFilterVersionCodes(currentVersionCodes googleapi.Int64s, newVersionCodes googleapi.Int64s) googleapi.Int64s {
+	sort.Slice(currentVersionCodes, func(a, b int) bool { return currentVersionCodes[a] < currentVersionCodes[b] })
+	sort.Slice(newVersionCodes, func(a, b int) bool { return newVersionCodes[a] < newVersionCodes[b] })
+
+	var filteredVersionCodes googleapi.Int64s
+	for i := 0; i < len(newVersionCodes); i++ {
+		log.Debugf("Searching for shadowing versions, comparing (%v) and (%v)", currentVersionCodes[i], newVersionCodes[i])
+		if currentVersionCodes[i] < newVersionCodes[i] {
+			log.Infof("Shadowing app found, removing current (%v) version, adding new (%v)", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, newVersionCodes[i])
+		} else {
+			log.Infof("Currently released app with (%v) version is higher than new (%v), app with new version code ignored", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, currentVersionCodes[i])
+		}
+	}
+	return filteredVersionCodes
+}
+
+// releaseStatusFromConfig gets the release status from the config value of user fraction.
+func releaseStatusFromConfig(userFraction float64) string {
+	if userFraction != 0 {
+		log.Infof("Release is a staged rollout, %v of users will receive it.", userFraction)
+		return releaseStatusInProgress
+	}
+	return releaseStatusCompleted
+}
+
+// getTrack gets the given track from the list of tracks of a given app.
+func getTrack(configs Configs, allTracks []*androidpublisher.Track) *androidpublisher.Track {
+	currentTrack := configs.Track
+	for _, track := range allTracks {
+		if currentTrack == track.Track {
+			log.Debugf("Current track found, name '%s'", currentTrack)
+			return track
+		}
+	}
+
+	return &androidpublisher.Track{
+		Releases:        []*androidpublisher.TrackRelease{},
+		Track:           currentTrack,
+		ServerResponse:  googleapi.ServerResponse{},
+		ForceSendFields: []string{},
+		NullFields:      []string{},
+	}
+}
+
+// getAllTracks lists all tracks for a given app.
+func getAllTracks(packageName string, service *androidpublisher.Service, appEdit *androidpublisher.AppEdit) ([]*androidpublisher.Track, error) {
```

Comment:
Would move logging out of this function.

Reply:
Decided to keep it, as it is used multiple times and would have to duplicate code.